contracts/misc/P2PPairStaking.sol

```diff
@@ -0,0 +1,618 @@
+        if (allowance == 0) {
+            IERC20(apeCoin).safeApprove(cApe, type(uint256).max);
+        }
+    }
+
+    function cancelListing(ListingOrder calldata listingOrder)
+        external
+        nonReentrant
+    {
+        require(msg.sender == listingOrder.offerer, "not order offerer");
+        bytes32 orderHash = getListingOrderHash(listingOrder);
+        require(!isListingOrderCanceled[orderHash], "order already cancel");
+        isListingOrderCanceled[orderHash] = true;
+
+        emit OrderCancelled(orderHash, listingOrder.offerer);
+    }
+
+    function matchPairStakingList(
+        ListingOrder calldata apeOrder,
+        ListingOrder calldata apeCoinOrder
+    ) external nonReentrant returns (bytes32 orderHash) {
+        //1 validate all order
+        _validateApeOrder(apeOrder);
+        _validateApeCoinOrder(apeCoinOrder);
+
+        //2 check if orders can match
+        require(
+            apeOrder.stakingType <= StakingType.MAYCStaking,
+            "invalid stake type"
+        );
+        require(
+            apeOrder.stakingType == apeCoinOrder.stakingType,
+            "orders type match failed"
+        );
+        require(
+            apeOrder.share + apeCoinOrder.share ==
+                PercentageMath.PERCENTAGE_FACTOR,
+            "orders share match failed"
+        );
+
+        //3 transfer token
+        IERC721(apeOrder.token).safeTransferFrom(
+            apeOrder.offerer,
+            address(this),
+            apeOrder.tokenId
+        );
+        uint256 apeAmount = getApeCoinStakingCap(apeOrder.stakingType);
+        IERC20(apeCoin).safeTransferFrom(
+            apeCoinOrder.offerer,
+            address(this),
+            apeAmount
+        );
+
+        //4 create match order
+        MatchedOrder memory matchedOrder = MatchedOrder({
+            stakingType: apeOrder.stakingType,
+            apeToken: apeOrder.token,
+            apeOfferer: apeOrder.offerer,
+            apeTokenId: apeOrder.tokenId,
+            apeShare: apeOrder.share,
+            bakcOfferer: address(0),
+            bakcTokenId: 0,
+            bakcShare: 0,
+            apeCoinOfferer: apeCoinOrder.offerer,
+            apeCoinShare: apeCoinOrder.share,
+            apePrincipleAmount: apeAmount
+        });
+        orderHash = getMatchedOrderHash(matchedOrder);
+        matchedOrders[orderHash] = matchedOrder;
+
+        //5 stake for ApeCoinStaking
+        ApeCoinStaking.SingleNft[]
+            memory singleNft = new ApeCoinStaking.SingleNft[](1);
+        singleNft[0].tokenId = apeOrder.tokenId;
+        singleNft[0].amount = apeAmount.toUint224();
+        if (apeOrder.stakingType == StakingType.BAYCStaking) {
+            apeCoinStaking.depositBAYC(singleNft);
+        } else {
+            apeCoinStaking.depositMAYC(singleNft);
+        }
+
+        //5 emit event
+        emit PairStakingMatched(orderHash);
+
+        return orderHash;
+    }
+
+    function matchBAKCPairStakingList(
```

Comment:
There's a lot of redundancy in the code between `matchPairStakingList` and `matchBAKCPairStakingList`. is it possible to abstract that redundancy?